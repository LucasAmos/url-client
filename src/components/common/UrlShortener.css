.url-shortener {
  margin-top: 70px;
  background-color: rgba(226, 226, 226, 0.2);
}

.url-shortener input {
  font-family: 'Roboto';
  font-weight: 100;
}

.url-shortener button {
  font-family: 'Roboto';
  font-weight: 100;
}

.url-shortener h2 {
padding-top: 25px;
font-size: 30px;
font-family: 'Roboto';
font-weight: 500;
}

.url-shortener .navbar-light .navbar-brand {
  color: #61dafb;
}

.url-shortener .navbar-light .navbar-nav .nav-link:hover, .navbar-light .navbar-nav .nav-link {
  color: rgb(189, 189, 189);
}

.url-shortener .navbar-light .navbar-nav .show > .nav-link {
  color: rgb(189, 189, 189);
}

.url-shortener .navbar-light .navbar-nav .nav-link:hover {
  color: #ffffff;
}

.url-shortener .row-one {
  margin-top: 50px;
  /* margin-bottom: 50px; */
  /* padding-bottom: 50px; */
}

.url-shortener .url-row {
  min-height: 60px;
}


.url-shortener .left-col {
  background-color: khaki;
  }

.url-shortener .right-col {
background-color: salmon;
}

.url-shortener .url {
  text-align: center;
  font-size: 24px;
}

.url-shortener .submit-button {
  width: 90px;
}

.url-shortener .submit-button.btn.btn-outline-secondary{
  border-top-right-radius: 0.25rem;
  border-bottom-right-radius: 0.25rem;
}

.url-shortener .submit-button.btn.btn-outline-secondary:hover {
  background-color: rgba(0, 0, 0, 0.1);
  color: #6c757d;
}

.url-shortener .submit-button.btn.btn-outline-secondary:active {
  border-color:  #ced4da;
}

.url-shortener .btn-outline-secondary {
  border: 1px solid #ced4da;
}


.url-shortener a {
  text-decoration: none;
}

.url-shortener a:hover {
  color: #013b79;
}

@media (max-width: 340px){
  .url-shortener span {
  max-width:320px;
  word-wrap:break-word;
  }
}

@media(min-width: 341px) and (max-width: 360px){
  .url-shortener span {
  max-width:340px;
  word-wrap:break-word;
  }
}

@media(min-width: 361px) and (max-width: 380px){
  .url-shortener span {
  max-width:360px;
  word-wrap:break-word;
  }
}

@media(min-width: 381px) and (max-width: 400px){
  .url-shortener span {
  max-width:380px;
  word-wrap:break-word;
  }
}

@media(min-width: 3px) and (max-width: 360px){
  .url-shortener span {
  max-width:340px;
  word-wrap:break-word;
  }
}

@media (max-width: 360px) {
  .url-shortener .row-one h2{
    font-size: 24px;
  }
}

@media (max-width: 400px) {
  .url-shortener .url-row {
    min-height: 80px;
  }

  @media (max-width: 500px) {
    .url-shortener {
      margin-top: 40px;
      margin-bottom: 40px;
    }
  }
}


