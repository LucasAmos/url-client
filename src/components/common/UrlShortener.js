/* eslint-disable react/jsx-one-expression-per-line */
import React, { Component } from 'react';
import validator from 'validator';
import Fade from 'react-reveal/Fade';
import withReveal from 'react-reveal/withReveal';
import { Container, Row, Col, Form, Button, InputGroup, FormControl } from 'react-bootstrap';
import config from '../../config/config';
import loadingImage from '../../res/images/loading.gif';
import './UrlShortener.css';

class UrlShortener extends Component {
  static formatUrl(url) {
    let result = url.replace(/^(?:https?:\/\/)?(?:www\.)?/i, '');
    result = `http://www.${result}`;
    return result;
  }

  constructor() {
    super();
    this.state = {
      loading: false,
      url: '',
      validUrl: true,
      link: '',
      shortenRoute: process.env.NODE_ENV === 'development' ? config.developmentURL : config.productionSubdomain,
      route: process.env.NODE_ENV === 'development' ? config.developmentURL : config.productionURL
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    const { route } = this.state;

    // ping heroku to activate the dyno
    // not nessary now on paid plan
    // fetch(`${route}status`)
    //   .then((response) => {
    //     response.json();
    //   });
  }

  handleSubmit() {
    const { url, route, shortenRoute } = this.state;
    const validURL = validator.isURL(url);

    this.setState({
      validUrl: validator.isURL(url)
    });

    if (validURL) {
      this.setState({ loading: true });
      fetch(`${route}api/shorten`, {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          url: UrlShortener.formatUrl(url)
        })
      })
        .then(response => response.json())
        .then(res => {
          this.setState({
            link: shortenRoute + res.hash,
            loading: false
          });
        })
        .catch(() => this.setState({ loading: false }));
    }
  }

  handleChange(e) {
    this.setState({ url: e.target.value });
  }

  render() {
    const { validUrl, url, link, loading } = this.state;
    return (
      <div className="url-shortener">
        <Container>
          <Row className="row-one">
            <Col className="col-one" lg={{ span: 8, offset: 2 }}>
              <h2>Need to shorten a URL?</h2>
            </Col>
            <Col className="col-two" lg={{ span: 8, offset: 2 }}>
              <Form.Group>
                <InputGroup>
                  <FormControl
                    onKeyDown={e => (e.key === 'Enter' ? this.handleSubmit() : null)}
                    className="url-input"
                    test="form-name"
                    name="name"
                    isInvalid={!validUrl}
                    type="text"
                    value={url}
                    onChange={e => {
                      this.setState({
                        url: e.target.value,
                        validUrl: e.target.value.length > 0
                      });
                    }}
                    placeholder="Enter a URL"
                  />
                  <InputGroup.Append>
                    <Button className="submit-button" variant="outline-secondary" onClick={this.handleSubmit}>
                      {' '}
                      {loading ? <img src={loadingImage} alt="loading" /> : 'Get URL!'}
                    </Button>
                  </InputGroup.Append>
                  <Form.Control.Feedback type="invalid">Please enter a valid URL</Form.Control.Feedback>
                </InputGroup>
              </Form.Group>
            </Col>
          </Row>
          <div className="url-row">
            {link && (
              <Row>
                <Col className="url" lg={{ span: 8, offset: 2 }}>
                  <span>
                    <a target="_blank" rel="noopener noreferrer" href={link}>
                      {link}
                    </a>
                  </span>
                </Col>
              </Row>
            )}
          </div>
        </Container>
      </div>
    );
  }
}

export default withReveal(UrlShortener, <Fade />);
