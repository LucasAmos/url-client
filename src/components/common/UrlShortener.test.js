import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import UrlShortener from './UrlShortener';

describe('UrlShortener component', () => {
  let savedError;
  beforeEach(() => {
    savedError = console.error;
    console.error = jest.fn();
  });

  afterEach(() => {
    console.error = savedError;
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      shallow(
        <UrlShortener />,
      ),
      div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it('matches the snapshot', () => {
    const tree = renderer.create(
      <UrlShortener />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
