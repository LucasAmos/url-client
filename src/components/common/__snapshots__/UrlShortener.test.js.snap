// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UrlShortener component matches the snapshot 1`] = `
<div
  className="react-reveal"
  style={
    Object {
      "opacity": undefined,
    }
  }
>
  <div
    className="url-shortener"
  >
    <div
      className="container"
    >
      <div
        className="row-one row"
      >
        <div
          className="col-one col-lg-8 offset-lg-2"
        >
          <h2>
            Need to shorten a URL?
          </h2>
        </div>
        <div
          className="col-two col-lg-8 offset-lg-2"
        >
          <div
            className="form-group"
          >
            <div
              className="input-group"
            >
              <input
                className="url-input form-control"
                name="name"
                onChange={[Function]}
                onKeyDown={[Function]}
                placeholder="Enter a URL"
                test="form-name"
                type="text"
                value=""
              />
              <div
                className="input-group-append"
              >
                <button
                  className="submit-button btn btn-outline-secondary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                    
                  Get URL!
                </button>
              </div>
              <div
                className="invalid-feedback"
              >
                Please enter a valid URL
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="url-row"
      >
        
      </div>
    </div>
  </div>
</div>
`;
